### DISCUSSION TITLE
What does the following commands do? 🤔

### DISCUSSION THREAD

**Discussion by limaceous-bushwhacker** [gokcehan/lf#1430] [Q&A] [answered]
- `set -f` - I see this being used a lot. Google tells me it's to disable filename expansions; is it supposed to be always used when writing functions?
- `set -- (cat ~/.local/share/lf/files)` - I've no idea what this one does. It's included in some of the examples in `Wiki/Tips` e.g. https://github.com/gokcehan/lf/wiki/Tips#create-symlinks-soft--hard

Since we're at it I'd like to throw in a couple more queries I had -
1. Is `$fx` supposed to be quoted or not? I see sometimes it being quotes in the aforementioned `Tips` section but sometimes not?
2. Can we use `\0` as as ifs?
3. How can I access the list of selected files outside lf? I'm aware of `~/.local/share/lf/files` but it isn't always populated with the list? Am I missing something or this warrant creating an issue?

Thanks in advance for your time and thanks to the contributors for working on this great project :)

**Comment by michaelferon** [comment, marked as answer]
`set` is used to modify shell options and set positional parameters. `set -f` disables globbing. Run
```bash
echo *
set -f
echo *
```
to see the effect.

`~/.local/share/lf/files` contains a list of files which have been marked for copying (yanking) or moving (cutting). The first line of this file indicates the mode (copy or move). Using `set` in this way is entirely distinct from the way it is used just above. It is simply setting the positional parameters, so that `$1` contains the mode of operation, and the remaining positional parameters contain the files marked for copying or moving.

1. `bash` splits parameters expansions on whitespace. It's considered good practice to quote parameter expansions to avoid unwanted field splitting.
2. Not sure about this as I haven't done this sorry :( but will look into it
3. You can access files which have been marked for copying or moving via `~/.local/share/lf/files`. This doesn't include files which have been selected using the `toggle` command (mapped to `<space>` by default).

**Reply by michaelferon** [reply]
There is some discussion on using \0 as a file separator in #47, but I'm not sure what the status is.

**Reply by michaelferon** [reply]
To elaborate a bit, I think the benefit in some of these instances of using `set -f` is so that `$fx` can be left unquoted. This makes using `$fx` a bit easier because, assuming none of the files contain spaces, each line of `$fx` will be treated as single 'word' (and hence, in the case of `set -- $(cat ~/.local/share/lf/files)`, each a single parameter). But if any of the files contain globbing characters, if left unquoted, they'll be interpreted as such. So disabling globbing in these instances is probably a smart thing to do.

**Reply by limaceous-bushwhacker** [reply]
Thank you so much for such an informative answer Michael. `set -- (cat ~/.local/share/lf/files)` appear to be a syntax to create arguments from within a function. I wasn't aware of this syntax before inspite of using *nix for years now.

From what I've seen looks like it's better to leave `$f, $fs, $fx` unquoted wherever possible. This throws me off since usually it's recommended to always quote the variables. I was curious why it's this way instead of just using NUL as a separator, that'd just remove all the confusion.

Now I'm curious about the difference between `toggle` and `select` 🤔

**Reply by michaelferon** [reply]
In general, parameters really should almost always be quoted, so your concern there is warranted. The examples in the wiki assume you have the following in your `lfrc`:
```bash
set ifs "\n"
set filesep "\n" # <- this is the default
```
As long as you have that set, you can safely leave `$f $fs $fx` unquoted. Without `ifs` set to `\n`, `bash` will split `$fx` on ANY whitespace it finds, and expansions of `$fx`, as well as `set -- $(cat ~/.local/share/lf/files)`, could behave unexpectedly (you could, for example, end up deleting a file you did not intend to delete).

I don't ever use `toggle` or `select` directly as commands. I just use the spacebar to "toggle" files (as you probably do as well), and `hjkl` keys to move around. As far as I can tell, the `select` command just takes an argument and moves the cursor line to that argument in the current directory.

P.S. You can always switch to `zsh`, where all these quoting difficulties can be handled much more elegantly :)

**Comment by superiums** [comment]
> `set` is used to modify shell options and set positional parameters. `set -f` disables globbing. Run
> 
> ```shell
> echo *
> set -f
> echo *
> ```
> 
> to see the effect.
> 
> `~/.local/share/lf/files` contains a list of files which have been marked for copying (yanking) or moving (cutting). The first line of this file indicates the mode (copy or move). Using `set` in this way is entirely distinct from the way it is used just above. It is simply setting the positional parameters, so that `$1` contains the mode of operation, and the remaining positional parameters contain the files marked for copying or moving.
> 
> 1. `bash` splits parameters expansions on whitespace. It's considered good practice to quote parameter expansions to avoid unwanted field splitting.
> 2. Not sure about this as I haven't done this sorry :( but will look into it
> 3. You can access files which have been marked for copying or moving via `~/.local/share/lf/files`. This doesn't include files which have been selected using the `toggle` command (mapped to `<space>` by default).

i tried the cmds , but they have different actions:
## in fish
`% echo *` : output is '*'
`% set -f` : nothing
`% echo *`: output  files in current dir.

## in bash
`% echo *` : output  files in current dir.
`% set -f` : nothing
`% echo *`: output  files in current dir.

are you sure `set -f` was used to disable glob ?
